tools/tarball_checker/check.py

```diff
@@ -0,0 +1,141 @@
+#!/usr/bin/env python3
+
+import argparse
+import sys
+import os
+import subprocess
+import tarfile
+import filecmp
+
+from urllib.request import urlretrieve
+import auditinfo
+
+# This script downloads and checks the source archive released by upstream.
+# When the current build targets an unreleased version, the script will
+# report an error and fail. Note that the GitHub workflow should make sure
+# that it is not executed in such cases.
+
+# Files that may be different without causing an error
+TO_IGNORE = [
+    'src/build-data/version.txt' # This contains specific release information, and may differ
+    ]
+
+# The maximum number of paths to print when differences are found
+MAX_DIFFERENCES = 10
+
+def _print_error(msg):
+    print(msg, file=sys.stderr)
+
+
+def _run(cmd, working_directory=".", stderr_ok=False):
+    p = subprocess.Popen(cmd, cwd=working_directory,
+                         stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    stdout, stderr = p.communicate()
+    errcode = p.wait()
+    if errcode != 0 or (stderr and not stderr_ok):
+        raise RuntimeError(f"Failed to run {cmd[0]}: \n{stderr}")
+    return stdout
+
+
+def download(url: str) -> str:
+    file, httpresponse = urlretrieve(url)
+    return file
+
+
+# Source: https://stackoverflow.com/a/24860799/4842497
+class dircmp(filecmp.dircmp):
+    """
+    Compare the content of dir1 and dir2. In contrast with filecmp.dircmp, this
+    subclass compares the content of files with the same path, not just the
+    file attributes.
+    """
+    def phase3(self):
+        fcomp = filecmp.cmpfiles(self.left, self.right, self.common_files,
+                                 shallow=False)
+        self.same_files, self.diff_files, self.funny_files = fcomp
+
+
+def recursive_diff(dir1, dir2):
+    compared = dircmp(dir1, dir2)
+    differing_paths = []
+    if (compared.left_only or compared.right_only or compared.diff_files
+        or compared.funny_files):
+        return compared.left_only + compared.right_only + compared.diff_files + compared.funny_files
+    for subdir in compared.common_dirs:
+        if diff_entries := recursive_diff(os.path.join(dir1, subdir), os.path.join(dir2, subdir)):
+            differing_paths += [os.path.join(subdir, entry) for entry in diff_entries]
```

Comment:
Don't we need to check that there aren't more (non-common) subdirectories than `compared.common_dirs` in either directory?

Reply:
The way I understand [the documentation](https://docs.python.org/3/library/filecmp.html), those show up in `left_only` and `right_only` and would therefore fail the comparison as is.

Reply:
Please merge, when you are convinced by this. :)

Reply:
Yeah, I think you are right!